File: backend/src/routes/store.js
const express = require("express");
const crypto = require("crypto");
const pool = require("../db");

const router = express.Router();

// --------------- Products (public) ---------------

// GET /api/store/products?store_type=ro|ups&category=&search=&featured=
router.get("/products", async (req, res) => {
  try {
    const { store_type, category, search, featured } = req.query;
    const conditions = [];
    const params = [];
    let idx = 1;

    if (store_type) {
      conditions.push(`store_type = $${idx++}`);
      params.push(store_type);
    }
    if (category) {
      conditions.push(`category = $${idx++}`);
      params.push(category);
    }
    if (search) {
      conditions.push(`(name ILIKE $${idx} OR description ILIKE $${idx} OR brand ILIKE $${idx})`);
      params.push(`%${search}%`);
      idx++;
    }
    if (featured === "true") {
      conditions.push("is_featured = true");
    }

    const where = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
    const result = await pool.query(
      `SELECT * FROM store_products ${where} ORDER BY is_featured DESC, id`,
      params
    );
    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to fetch products" });
  }
});

// GET /api/store/products/:id
router.get("/products/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const [product, reviews] = await Promise.all([
      pool.query("SELECT * FROM store_products WHERE id = $1", [id]),
      pool.query("SELECT * FROM store_reviews WHERE product_id = $1 ORDER BY created_at DESC", [id]),
    ]);
    if (product.rows.length === 0) return res.status(404).json({ error: "Product not found" });
    res.json({ ...product.rows[0], reviews: reviews.rows });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to fetch product" });
  }
});

// GET /api/store/categories?store_type=ro|ups
router.get("/categories", async (req, res) => {
  try {
    const { store_type } = req.query;
    const result = await pool.query(
      "SELECT DISTINCT category FROM store_products WHERE store_type = $1 ORDER BY category",
      [store_type || "ro"]
    );
    res.json(result.rows.map((r) => r.category));
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to fetch categories" });
  }
});

// --------------- Cart (session-based) ---------------

// GET /api/store/cart?session_id=xxx
router.get("/cart", async (req, res) => {
  try {
    const { session_id } = req.query;
    if (!session_id) return res.status(400).json({ error: "session_id required" });

    const result = await pool.query(
      `SELECT sc.id, sc.qty, sp.*
       FROM store_cart sc JOIN store_products sp ON sc.product_id = sp.id
       WHERE sc.session_id = $1 ORDER BY sc.id`,
      [session_id]
    );
    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to fetch cart" });
  }
});

// POST /api/store/cart
router.post("/cart", async (req, res) => {
  try {
    const { session_id, product_id, qty } = req.body;
    if (!session_id || !product_id) return res.status(400).json({ error: "session_id and product_id required" });

    // Check if item already in cart
    const existing = await pool.query(
      "SELECT * FROM store_cart WHERE session_id = $1 AND product_id = $2",
      [session_id, product_id]
    );

    if (existing.rows.length > 0) {
      const updated = await pool.query(
        "UPDATE store_cart SET qty = qty + $1 WHERE id = $2 RETURNING *",
        [qty || 1, existing.rows[0].id]
      );
      return res.json(updated.rows[0]);
    }

    const result = await pool.query(
      "INSERT INTO store_cart (session_id, product_id, qty) VALUES ($1, $2, $3) RETURNING *",
      [session_id, product_id, qty || 1]
    );
    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to add to cart" });
  }
});

// PUT /api/store/cart/:id
router.put("/cart/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const { qty } = req.body;
    if (qty <= 0) {
      await pool.query("DELETE FROM store_cart WHERE id = $1", [id]);
      return res.json({ message: "Removed" });
    }
    const result = await pool.query("UPDATE store_cart SET qty = $1 WHERE id = $2 RETURNING *", [qty, id]);
    res.json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to update cart" });
  }
});

// DELETE /api/store/cart/:id
router.delete("/cart/:id", async (req, res) => {
  try {
    await pool.query("DELETE FROM store_cart WHERE id = $1", [req.params.id]);
    res.json({ message: "Removed" });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to remove from cart" });
  }
});

// --------------- Reviews ---------------

// POST /api/store/reviews
router.post("/reviews", async (req, res) => {
  try {
    const { product_id, author, rating, comment } = req.body;
    if (!product_id || !rating) return res.status(400).json({ error: "product_id and rating required" });

    const result = await pool.query(
      "INSERT INTO store_reviews (product_id, author, rating, comment) VALUES ($1, $2, $3, $4) RETURNING *",
      [product_id, author || "Anonymous", rating, comment || ""]
    );
    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to add review" });
  }
});

module.exports = router;

File: backend/app/routers/store.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from typing import Optional

from ..database import get_db
from ..models import StoreProducts, StoreCart, StoreReviews
from ..schemas import (
    StoreProductOut,
    StoreProductDetail,
    StoreReviewCreate,
    StoreReviewOut,
    CartItemCreate,
    CartItemUpdate,
    CartItemOut,
    MessageResponse,
)

router = APIRouter(prefix="/api/store", tags=["store"])


# ─── Products ───
@router.get("/products", response_model=list[StoreProductOut])
def list_products(
    store_type: Optional[str] = Query(None),
    category: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    featured: Optional[str] = Query(None),
    db: Session = Depends(get_db),
):
    q = db.query(StoreProducts)
    conditions = []

    if store_type:
        conditions.append(StoreProducts.store_type == store_type)
    if category:
        conditions.append(StoreProducts.category == category)
    if search:
        like = f"%{search}%"
        conditions.append(
            (StoreProducts.name.ilike(like))
            | (StoreProducts.description.ilike(like))
            | (StoreProducts.brand.ilike(like))
        )
    if featured == "true":
        conditions.append(StoreProducts.is_featured == True)

    if conditions:
        from sqlalchemy import and_

        q = q.filter(and_(*conditions))

    return q.order_by(StoreProducts.is_featured.desc(), StoreProducts.id).all()


@router.get("/products/{id}", response_model=StoreProductDetail)
def get_product(id: int, db: Session = Depends(get_db)):
    product = db.query(StoreProducts).filter(StoreProducts.id == id).first()
    if not product:
        raise HTTPException(status_code=404, detail="Product not found")
    reviews = (
        db.query(StoreReviews)
        .filter(StoreReviews.product_id == id)
        .order_by(StoreReviews.created_at.desc())
        .all()
    )
    result = StoreProductDetail.model_validate(product)
    result.reviews = [StoreReviewOut.model_validate(r) for r in reviews]
    return result


@router.get("/categories", response_model=list[str])
def list_categories(
    store_type: Optional[str] = Query("ro"),
    db: Session = Depends(get_db),
):
    rows = (
        db.query(StoreProducts.category)
        .filter(StoreProducts.store_type == store_type)
        .distinct()
        .order_by(StoreProducts.category)
        .all()
    )
    return [r[0] for r in rows]


# ─── Cart ───
@router.get("/cart", response_model=list[CartItemOut])
def get_cart(
    session_id: str = Query(...),
    db: Session = Depends(get_db),
):
    rows = (
        db.query(StoreCart, StoreProducts)
        .join(StoreProducts, StoreCart.product_id == StoreProducts.id)
        .filter(StoreCart.session_id == session_id)
        .order_by(StoreCart.id)
        .all()
    )
    result = []
    for cart_item, product in rows:
        item = CartItemOut(
            id=cart_item.id,
            session_id=cart_item.session_id,
            product_id=cart_item.product_id,
            qty=cart_item.qty,
            name=product.name,
            brand=product.brand,
            category=product.category,
            store_type=product.store_type,
            price=float(product.price) if product.price else None,
            images=product.images,
            specs=product.specs,
            description=product.description,
            features=product.features,
            warranty=product.warranty,
            is_featured=product.is_featured,
            created_at=cart_item.created_at,
        )
        result.append(item)
    return result


@router.post("/cart", status_code=201)
def add_to_cart(
    body: CartItemCreate,
    db: Session = Depends(get_db),
):
    existing = (
        db.query(StoreCart)
        .filter(
            StoreCart.session_id == body.session_id,
            StoreCart.product_id == body.product_id,
        )
        .first()
    )
    if existing:
        existing.qty += body.qty or 1
        db.commit()
        db.refresh(existing)
        return {
            "id": existing.id,
            "session_id": existing.session_id,
            "product_id": existing.product_id,
            "qty": existing.qty,
        }

    cart_item = StoreCart(
        session_id=body.session_id,
        product_id=body.product_id,
        qty=body.qty or 1,
    )
    db.add(cart_item)
    db.commit()
    db.refresh(cart_item)
    return {
        "id": cart_item.id,
        "session_id": cart_item.session_id,
        "product_id": cart_item.product_id,
        "qty": cart_item.qty,
    }


@router.put("/cart/{id}")
def update_cart(
    id: int,
    body: CartItemUpdate,
    db: Session = Depends(get_db),
):
    if body.qty <= 0:
        db.query(StoreCart).filter(StoreCart.id == id).delete()
        db.commit()
        return {"message": "Removed"}

    cart_item = db.query(StoreCart).filter(StoreCart.id == id).first()
    if not cart_item:
        raise HTTPException(status_code=404, detail="Cart item not found")
    cart_item.qty = body.qty
    db.commit()
    db.refresh(cart_item)
    return {
        "id": cart_item.id,
        "session_id": cart_item.session_id,
        "product_id": cart_item.product_id,
        "qty": cart_item.qty,
    }


@router.delete("/cart/{id}", response_model=MessageResponse)
def remove_from_cart(id: int, db: Session = Depends(get_db)):
    db.query(StoreCart).filter(StoreCart.id == id).delete()
    db.commit()
    return {"message": "Removed"}


# ─── Reviews ───
@router.post("/reviews", response_model=StoreReviewOut, status_code=201)
def add_review(body: StoreReviewCreate, db: Session = Depends(get_db)):
    review = StoreReviews(**body.model_dump())
    db.add(review)
    db.commit()
    db.refresh(review)
    return review
